src/core/unit.c

```diff
@@ -481,6 +481,9 @@ bool unit_may_gc(Unit *u) {
                         return false;
         }
 
+        if (u->in_cgroup_empty_queue)
+                return false;
```

Comment:
this is a relatively cheap check (i.e. doesn't go to the vfs, doesn't check any other objects), hence please put it somewhere early in the function, maybe right after the check if the unit is perpetual.

otherwise we might do all the expensive work for nothing, and then decide it was all useless after checking the final cheap check...